plugins/logitech-tap/fu-logitech-tap-sensor-device.c

```diff
@@ -0,0 +1,390 @@
+				  data,
+				  NULL,
+				  FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+				  &error_local)) {
+		g_debug("failed to send get request, retrying: %s", error_local->message);
+		if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+					  HIDIOCGINPUT(datasz),
+					  data,
+					  NULL,
+					  FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+					  error))
+			return FALSE;
+	}
+
+	fu_dump_raw(G_LOG_DOMAIN, "HidGetFeatureRes", data, datasz);
+
+	return TRUE;
+#else
+	/* failed */
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_enable_tde(FuDevice *device, GError **error)
+{
+	FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	guint8 set_data[HID_SET_DATA_LEN] = {kHidMcuTdeReportId,
+					     kHidMcuTdeModeSelector,
+					     kHidMcuTdeModeEnable,
+					     0,
+					     0};
+
+	/* enable/disable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						       (guchar *)set_data,
+						       HID_SET_DATA_LEN,
+						       error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_disable_tde(FuDevice *device, GError **error)
+{
+	FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	guint8 set_data[HID_SET_DATA_LEN] = {kHidMcuTdeReportId,
+					     kHidMcuTdeModeSelector,
+					     kHidMcuTdeModeDisable,
+					     0,
+					     0};
+
+	/* enable/disable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						       (guchar *)set_data,
+						       HID_SET_DATA_LEN,
+						       error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_logitech_tap_sensor_device_reboot_device(FuDevice *device, GError **error)
+{
+	FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	gint8 pinclr = 05;
+	guint8 pinset = 06;
+	guint8 PWR = 45;
+	guint8 RST = 46;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	guint8 set_data[HID_SET_DATA_LEN] = {kHidReportIdMcuSetCmd, pinclr, PWR, 0, 0};
+	g_autoptr(FuProgress) progress = fu_progress_new(G_STRLOC);
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 100, "attach");
+	fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_RESTART);
+
+	/* need to reopen the device, as at composite_cleaup time, device is already closed */
+	if (!fu_device_open(FU_DEVICE(device), error))
+		return FALSE;
+
+	/* enable/disable TDE mode */
+	locker =
+	    fu_device_locker_new_full(device,
+				      (FuDeviceLockerFunc)fu_logitech_tap_sensor_device_enable_tde,
+				      (FuDeviceLockerFunc)fu_logitech_tap_sensor_device_disable_tde,
+				      error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* setup HID report for power cycle */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						       (guchar *)set_data,
```

Comment:
guint8 would be more normal for byte arrays

Reply:
Fixed